src/groupeddataframe/splitapplycombine.jl

```diff
@@ -693,22 +563,89 @@ end
-                                   "is not equal to grouping key :$key"))
+            if hasproperty(valscat, key)
+                if keeprows
+                    isequal(valscat[!, key], parent(gd)[!, key]) ||
+                    throw(ArgumentError("column :$key in returned data frame " *
+                                        "is not equal to grouping key :$key"))
+
+                else
+                    isequal(valscat[!, key], view(parent(gd)[!, key], idx)) ||
+                    throw(ArgumentError("column :$key in returned data frame " *
+                                        "is not equal to grouping key :$key"))
+                end
             end
         end
-        return hcat!(parent(gd)[idx, gd.cols],
-                     select(valscat, Not(intersect(keys, _names(valscat))), copycols=false))
+        if keeprows
+            newparent = hcat!(select(parent(gd), gd.cols, copycols=copycols),
+                              select(valscat, Not(intersect(keys, _names(valscat))),
+                                     copycols=false), copycols=false)
+        else
+            newparent = hcat!(parent(gd)[idx, gd.cols],
+                              select(valscat, Not(intersect(keys, _names(valscat))),
+                                     copycols=false), copycols=false)
+        end
+        regroup || return newparent
+
+        if length(idx) == 0
+            @assert nrow(newparent) == 0
+            return GroupedDataFrame(newparent, collect(1:length(gd.cols)), Int[],
+                                    Int[], Int[], Int[], 0, Dict{Any,Int}())
+        end
+        if keeprows
+            # in this case we are sure that the result GroupedDataFrame has the
+            # same structure as the source
+            # we do not copy data as it should be safe - we never mutate fields of gd
+            return GroupedDataFrame(newparent, gd.cols, gd.groups, gd.idx,
+                                    gd.starts, gd.ends, gd.ngroups, getfield(gd, :keymap))
+        else
+            starts = Vector{Int}(undef, length(gd))
+            ends = Vector{Int}(undef, length(gd))
+            starts[1] = 1
+            j = 1
+            for i in 2:length(idx)
+                if idx[i] != idx[i-1]
+                    j += 1
+                    starts[j] = i
+                    ends[j-1] = i - 1
+                end
+            end
+            # it is impossible to get more groups in the output than we had initially
+            @assert j <= length(gd)
+            # In case some groups have to be dropped
+            resize!(starts, j)
+            resize!(ends, j)
+            ends[end] = length(idx)
+
+            groups = zeros(Int, length(idx))
+            for i in 1:j
+                @inbounds for k in starts[i]:ends[i]
+                    groups[k] = i
+                end
+            end
+            # all groups must be filled
+            @assert minimum(groups) == 1
```

Comment:
I guess this is what this is supposed to test?
```suggestion
            @assert !any(iszero, groups)
```

Reply:
your proposal is a bit slower (I guess `minimum` is able to avoid branching in the loop). Additionally my proposal additionally checks for errors (which we had in the past) in case `groups` is invalid and does not start from `1`. Therefore I leave my implementation.